build-support: Add fetchfossil function.

MostAwesomeDude opened the pull request:
###### Motivation for this change

I have a Fossil repository that I want to reference from Nix.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This is a dead-simple fetcher which clones a Fossil repository, opens it directly into $out, and then nicks out the single Fossil checkout marker.

For some reason, my Let's Encrypt certificates don't work when fetching over HTTPS. I don't know why. It's not a dealbreaker for me, though.

MostAwesomeDude committed:
build-support: Add fetchfossil function.

This is a dead-simple fetcher which clones a Fossil repository, opens it
directly into $out, and then nicks out the single Fossil checkout marker.

MostAwesomeDude committed:
fetchfossil: Don't let Fossil write to $out/.fossil.

garbas commented on pkgs/build-support/fetchfossil/default.nix:
```diff
@@ -0,0 +1,21 @@
+
+{name ? null, url, rev ? null, md5 ? null, sha256 ? null}:
+
+# TODO: statically check if mercurial as the https support if the url starts woth https.
```
i guess this is a leftover from fetchhg right?

MostAwesomeDude replied:
Ha, yeah. Lemme fix that up.

MostAwesomeDude committed:
fetchfossil: Cleanup some copy-and-paste.

MostAwesomeDude committed:
fetchfossil: Require rev.

Otherwise no bumps are done when upstream changes. Not sure how to ping the
repository and check without just doing the clone.

MostAwesomeDude commented:
This is ready for review; I've been using this for a couple days without incident.

domenkozar commented on pkgs/build-support/fetchfossil/default.nix:
```diff
@@ -0,0 +1,20 @@
+{stdenv, fossil}:
+
+{name ? null, url, rev, md5 ? null, sha256 ? null}:
```
We shouldn't support `md5` in newer functions. We're trying to get rid of it. See https://github.com/NixOS/nixpkgs/issues/4491

MostAwesomeDude replied:
And good riddance! Changing it to mandate SHA256.

MostAwesomeDude committed:
fetchfossil: Don't support MD5.

This effectively means that SHA256 is the only way to go. I'm not really
bothered by this.

domenkozar commented:
Looks good. Is there a `.git` equivalent that would change through time (non-deterministic)?

MostAwesomeDude commented:
No, Fossil repositories are a single SQLite database. The `fossil open` command generates a working tree from the database, and there's just the single `.fslckout` file left in there.
